File: dupe/dupe.go
package main

import (
	"crypto/md5"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
)

var verbose *bool = flag.Bool("verbose", false, "Print the list of duplicate files.")
var rootDir string = "."
var fullPathsByFilename map[string][]string

func VisitFile(fullpath string, f os.FileInfo, err error) error {
    if err != nil {
        return err
    }
    if !f.IsDir() {
        filename := path.Base(fullpath)
        fullPathsByFilename[filename] = append(fullPathsByFilename[filename], fullpath)
    }
    return nil;
}

func MD5OfFile(fullpath string) []byte {
	if contents, err := ioutil.ReadFile(fullpath); err == nil {
		md5sum := md5.New()
		md5sum.Write(contents)
		return md5sum.Sum(nil)
	}
	return nil
}

func PrintResults() {
	dupes := 0
	for key, value := range fullPathsByFilename {
		if len(value) < 2 {
			continue
		}
		dupes++
		if *verbose {
			println(key, ":")
			for _, filename := range value {
				println("  ", filename)
				fmt.Printf("    %x\n", MD5OfFile(filename))
			}
		}
	}
	println("Total duped files found:", dupes)
}

func FindDupes(root string) {
	fullPathsByFilename = make(map[string][]string)
	filepath.Walk(root, VisitFile)
}

func ParseArgs() {
	flag.Parse()
	if len(flag.Args()) > 0 {
		rootDir = flag.Arg(0)
	}
}

func main() {
	ParseArgs()
	FindDupes(rootDir)
	PrintResults()
}

File: dupe/dupe.rb
require 'find'
require 'digest/md5'

VERBOSE = ARGV.delete('--verbose')
ROOT_DIR = ARGV[0] ||= "."

@full_paths_by_filename = {}

def md5_of_file(fullpath)
  Digest::MD5.hexdigest(File.read(fullpath))
end

def visit_file(fullpath)
  return if File.directory? fullpath
  return if File.symlink? fullpath
  basename = File.basename(fullpath)
  (@full_paths_by_filename[basename] ||= []) << fullpath
end

def print_results
  @full_paths_by_filename.select! { |filename, fullpaths| fullpaths.length >= 2 }
  if (VERBOSE)
    @full_paths_by_filename.each do |filename, fullpaths|
      puts "#{filename}:"
      for path in fullpaths do
        puts "  #{path}"
        puts "    #{md5_of_file(path)}"
      end
    end
  end
  puts "Total duped files found: #{@full_paths_by_filename.length}"
end

Find.find(ROOT_DIR) do |fullpath|
  visit_file(fullpath)
end

print_results